compiler/rustc_metadata/src/dependency_format.rs

```diff
@@ -156,25 +156,46 @@ fn calculate_type(tcx: TyCtxt<'_>, ty: CrateType) -> DependencyList {
         Linkage::Dynamic | Linkage::IncludedFromDylib => {}
     }
 
+    let all_dylibs = || {
+        tcx.crates(()).iter().filter(|&&cnum| {
+            !tcx.dep_kind(cnum).macros_only() && tcx.used_crate_source(cnum).dylib.is_some()
+        })
+    };
+
+    let mut upstream_in_dylibs = FxHashSet::default();
+
+    if sess.opts.unstable_opts.prefer_deps_of_dynamic || tcx.features().rustc_private {
+        // Find all libraries statically linked to upstream dylibs.
+        for &cnum in all_dylibs() {
+            let deps = tcx.dylib_dependency_formats(cnum);
+            for &(depnum, style) in deps.iter() {
+                if let RequireStatic = style {
+                    upstream_in_dylibs.insert(depnum);
+                }
+            }
+        }
+    }
```

Comment:
Instead of all this interspersed logic would it be possible to at the end iterate over all dylibs and mark all crates they have linked into them as `IncludedFromDylib`?

Reply:
Not sure if that's correct. The current logic starts by picking the dylibs then adding other crates as needed. That seems less likely to end up with missing crates than removing them after.

Reply:
That's what I mean by remove as `IncludeFromDylib` won't be linked to the crate. It also won't work with the current duplicate crate detection as that happens as the crates are added.